homeassistant/components/nest/sensor_sdm.py

```diff
@@ -0,0 +1,169 @@
+        if HumidityTrait.NAME in device.traits:
+            entities.append(HumiditySensor(device))
+    async_add_entities(entities)
+
+
+class SensorBase(Entity):
+    """Representation of a dynamically updated Sensor."""
+
+    def __init__(self, device: Device):
+        """Initialize the sensor."""
+        self._device = device
+
+    @property
+    def should_pool(self) -> bool:
+        """Disable polling since entities have state pushed via pubsub."""
+        return False
+
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID."""
+        # The API "name" field is a unique device identifier.
+        return f"{self._device.name}-{self.device_class}"
+
+    @property
+    def device_name(self):
+        """Return the name of the physical device that includes the sensor."""
+        if InfoTrait.NAME in self._device.traits:
+            trait = self._device.traits[InfoTrait.NAME]
+            if trait.custom_name:
+                return trait.custom_name
+        # Build a name from the room/structure.  Note: This room/structure name
+        # is not associated with a home assistant Area.
+        parent_relations = self._device.parent_relations
+        if parent_relations:
+            items = sorted(parent_relations.items())
+            names = [name for id, name in items]
+            return " ".join(names)
+        return self.unique_id
+
+    @property
+    def device_info(self):
+        """Return device specific attributes."""
+        return {
+            "identifiers": {(DOMAIN, self._device.name)},
+            "name": self.device_name,
+            "manufacturer": "Google Nest",
+            "model": self.device_model,
+        }
+
+    @property
+    def device_model(self):
+        """Return device model information."""
+        # The API intentionally returns minimal information about specific
+        # devices, instead relying on traits, but we can infer a generic model
+        # name based on the type
+        if self._device.type == "sdm.devices.types.CAMERA":
+            return "Camera"
+        if self._device.type == "sdm.devices.types.DISPLAY":
+            return "Display"
+        if self._device.type == "sdm.devices.types.DOORBELL":
+            return "Doorbell"
+        if self._device.type == "sdm.devices.types.THERMOSTAT":
+            return "Thermostat"
+        return None
+
+    async def async_added_to_hass(self):
+        """Run when entity is added to register update signal handler."""
+
+        async def async_update_state():
+            """Update sensor state."""
+            await self.async_update_ha_state(True)
+
+        # Event messages trigger the SIGNAL_NEST_UPDATE, which is intercepted
+        # here to re-fresh the signals from _device.  Unregister this callback
+        # when the entity is removed.
+        self.async_on_remove(
+            async_dispatcher_connect(self.hass, SIGNAL_NEST_UPDATE, async_update_state)
+        )
```

Comment:
This will call `await self.async_update()` and then `self.async_write_ha_state()`. Is there state to be updated or can we just run `self.async_write_ha_state()` directly inside a callback? 
```suggestion
        # Event messages trigger the SIGNAL_NEST_UPDATE, which is intercepted
        # here to re-fresh the signals from _device.  Unregister this callback
        # when the entity is removed.
        self.async_on_remove(
            async_dispatcher_connect(self.hass, SIGNAL_NEST_UPDATE, self.async_write_ha_state)
        )
```

Reply:
The _device itself should be updated by the time we get here.  This looks good to me.

Reply:
I tried this out and it ended up failing tests.  I'm not totally sure I understand what could be improved here.

By the time this runs, _device has been updated with the latest state by the DeviceManager in the pubsub callback.  Is this a threadsafety issue here? Python threading model isn't totally my specialty.